core/src/test/java/org/infinispan/persistence/file/StateTransferExpiredFileStoreTest.java

```diff
@@ -0,0 +1,85 @@
+package org.infinispan.persistence.file;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.infinispan.commons.test.CommonsTestingUtil.tmpDirectory;
+import static org.testng.AssertJUnit.assertEquals;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.time.ControlledTimeService;
+import org.infinispan.commons.time.TimeService;
+import org.infinispan.commons.util.Util;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestDataSCI;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.data.Address;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Reproducer for ISPN-13191 Deadlock: expiration blocks state transfer
+ */
+@Test(testName = "statetransfer.StateTransferExpiredFileStoreTest", groups = "functional")
+public class StateTransferExpiredFileStoreTest extends MultipleCacheManagersTest {
+   public static final String CACHE_NAME = "cache";
+   private ControlledTimeService timeService;
+   private int storeId;
+
+   @Override
+   public Object[] factory() {
+      return new Object[]{
+         new StateTransferExpiredFileStoreTest().cacheMode(CacheMode.REPL_SYNC)
```

Comment:
I copied that cache mode (and some other parts) from StateTransferTimestampsTest.java to mimic our WildFly setup and removed the other modes to make the testing less slow while trying to reproduce the bug. Putting those back may not be a bad idea.

Can we safely use this fix with12.1.x and use that infinispan-core with WildFly 24 despite the failing Jenkins here?

Thanks.

Reply:
Yeah, adding scattered mode was a good idea, because it turns out the fix introduces a new deadlock if the cache is scattered. AFAIK Hibernate 2LC uses scattered mode, so using the current code is not an option.